overlord/devicestate/remodel_test.go

```diff
@@ -0,0 +1,704 @@
+		"revision":       "1",
+	})
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	remodCtx, err := devicestate.RemodelCtx(s.state, oldModel, newModel)
+	c.Assert(err, IsNil)
+
+	c.Check(remodCtx.ForRemodeling(), Equals, true)
+
+	chg := s.state.NewChange("remodel", "...")
+
+	err = remodCtx.Init(chg)
+	c.Assert(err, IsNil)
+
+	var encNewModel string
+	c.Assert(chg.Get("new-model", &encNewModel), IsNil)
+
+	c.Check(encNewModel, Equals, string(asserts.Encode(newModel)))
+
+	c.Check(remodCtx.Model(), DeepEquals, newModel)
+	// an update remodel does not need a new/dedicated store
+	c.Check(remodCtx.Store(), IsNil)
+}
+
+func (s *remodelLogicSuite) TestNewStoreRemodelContextInit(c *C) {
+	oldModel := fakeRemodelingModel(nil)
+	newModel := fakeRemodelingModel(map[string]interface{}{
+		"store":    "my-other-store",
+		"revision": "1",
+	})
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// we have a device state
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:           "my-brand",
+		Model:           "my-model",
+		Serial:          "serialserialserial",
+		SessionMacaroon: "prev-session",
+	})
+
+	remodCtx, err := devicestate.RemodelCtx(s.state, oldModel, newModel)
+	c.Assert(err, IsNil)
+
+	c.Check(remodCtx.ForRemodeling(), Equals, true)
+
+	chg := s.state.NewChange("remodel", "...")
+
+	err = remodCtx.Init(chg)
+	c.Assert(err, IsNil)
+
+	var encNewModel string
+	c.Assert(chg.Get("new-model", &encNewModel), IsNil)
+
+	c.Check(encNewModel, Equals, string(asserts.Encode(newModel)))
+
+	var device *auth.DeviceState
+	c.Assert(chg.Get("device", &device), IsNil)
+	// session macaroon was reset
+	c.Check(device, DeepEquals, &auth.DeviceState{
+		Brand:  "my-brand",
+		Model:  "my-model",
+		Serial: "serialserialserial",
+	})
+
+	c.Check(remodCtx.Model(), DeepEquals, newModel)
+}
+
+func (s *remodelLogicSuite) TestRemodelDeviceBackendNoChangeYet(c *C) {
+	oldModel := fakeRemodelingModel(nil)
+	newModel := fakeRemodelingModel(map[string]interface{}{
+		"store":    "my-other-store",
+		"revision": "1",
+	})
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// we have a device state
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:  "my-brand",
+		Model:  "my-model",
+		Serial: "serialserialserial",
+	})
+
+	remodCtx, err := devicestate.RemodelCtx(s.state, oldModel, newModel)
+	c.Assert(err, IsNil)
+
+	devBE := devicestate.RemodelDeviceBackend(remodCtx)
+
+	device, err := devBE.Device()
+	c.Assert(err, IsNil)
+	c.Check(device, DeepEquals, &auth.DeviceState{
+		Brand:  "my-brand",
+		Model:  "my-model",
+		Serial: "serialserialserial",
+	})
+
+	mod, err := devBE.Model()
+	c.Assert(err, IsNil)
+	c.Check(mod, DeepEquals, newModel)
+
+	err = devBE.SetDevice(&auth.DeviceState{
+		Brand:           "my-brand",
+		Model:           "my-model",
+		Serial:          "serialserialserial",
+		SessionMacaroon: "session",
+	})
+	c.Assert(err, IsNil)
+
+	expectedDevice := &auth.DeviceState{
```

Comment:
The call to "SetDevice()" above sets to exactly the same device-state, should this just moved up and devBE.SetDevice(expectedDevice) be called ? Or is that confusing?